feat(integrations/sendgrid): Add integration for the SendGrid Email API

JustinBordage opened the pull request.

JustinBordage committed:
chore(sendgrid): Add JetBrains IDE metadata files to gitignore

JustinBordage committed:
Merge branch 'master' into jbord_sendgrid-integration

JustinBordage committed:
feat(sendgrid): Add implementation for SendGrid Integration

JustinBordage committed:
Refactor(sendgrid): Removed duplicate code I forgot to remove

JustinBordage committed:
Chore(SendGrid): Fix linter issues

franklevasseur commented on integrations/sendgrid/src/misc/validation-utils.ts:
```diff
@@ -0,0 +1,30 @@
+import { RuntimeError } from '@botpress/sdk'
+import { ValidationResult } from './custom-types'
+
+export const validateData = <Schema extends ZodType>(schema: Schema, data: input<Schema>): ValidationResult<Schema> => {
```
**chore:** zod is such a core and important lib, that we try not to add an abstraction layer on top of it. This function is basically just a `.safeParse` with some error mapping; I think we can get rid of it

franklevasseur commented on integrations/sendgrid/src/misc/utils.ts:
```diff
@@ -0,0 +1,43 @@
+  return error && error instanceof ZodError
+}
+
+export const parseError = (thrown: any) => {
```
**chore:** You should use type `unknown` instead of `any`. unknown is basically the union of all possible types in TypeScript, while any just skips the type-checking altogether. Long story short; except when coding in the SDK, we should never use `any`

franklevasseur commented on integrations/sendgrid/src/misc/utils.ts:
```diff
@@ -0,0 +1,43 @@
+  sgMail.setApiKey(config.apiKey)
+}
+
+export const isSendGridError = (thrown: any): thrown is SendGridResponseError => {
```
**praise:** nice type guard

franklevasseur commented on integrations/sendgrid/src/misc/utils.ts:
```diff
@@ -0,0 +1,43 @@
+import { SendGridResponseError } from './custom-types'
+import { zodErrorToRuntimeError } from './validation-utils'
+
+export const initializeSendGrid = (config: bp.configuration.Configuration) => {
```
**chore:** Your register function should also ensure that the provided configuration is valid. This means you should probably try a dummy API call to SendGrid just to make sure it's successful

franklevasseur commented on integrations/sendgrid/src/misc/custom-schemas.ts:
**chore:** when we need to define schemas that are used both in the integration definition and implementation, we put them in `integrations/$YOUR_INTEGRATION/definitions` directory. See [gmail](https://github.com/botpress/botpress/tree/master/integrations/gmail/definitions) for reference.

JustinBordage replied:
chore: I moved the schemas to the definitions directory. Could you check if the change follows your expectations?

franklevasseur reviewed (commented):
Good work so far!! 

Here's a small round of comments. I'll do another round once the PR is ready for review (not draft).

We internally use [conventional comments](https://conventionalcomments.org/) to review PRs. Check it out when you have some time.

JustinBordage commented on integrations/sendgrid/src/index.ts:
```diff
@@ -0,0 +1,13 @@
+
+export default new bp.Integration({
+  register: async (props) => {
+    initializeSendGrid(props.ctx.configuration)
```
Do you mean get rid of the 'initializeSendGrid' function and just write `sgMail.setApiKey(props.ctx.configuration.apiKey)` instead?

If so, I don't really have a reason aside from trying to compartmentalize it similar to the "getMailchimpClient" function in the Mailchimp integration.

I'll change it to match the above snippet since it makes more sense anyway.

JustinBordage committed:
refactor: Move "initializeSendGrid" logic into Botpress register function to remove abstraction

JustinBordage committed:
fix: Fixed issue where incorrect schema was used for the action output

JustinBordage committed:
refactor: Removed unnecessary validation of output data for sendMail action

JustinBordage commented on integrations/sendgrid/src/actions/send-mail.ts:
```diff
@@ -0,0 +1,28 @@
+  try {
+    const [response, _] = await sgMail.send(validatedInput)
+
+    return sendEmailOutputSchema.parse({
```
This actually made me realize there was a bug in the integration definition file, since I wasn't using the correct schema there.

JustinBordage committed:
refactor: Removed unnecessary validation of input data for sendMail action

- Also removed 'validateData' function since it added unnecessary layer of abstraction.

JustinBordage committed:
refactor: Removed ZodError parsing from the parseError

This was done since both of the manual validations of input/output data within the sendMail action handler were removed. Therefore this will never be triggered anymore since the validation is handled by the Botpress backend

JustinBordage committed:
chore: Fixed usage of 'any' to instead be 'unknown'

JustinBordage committed:
docs: fixed incorrect usage of word 'request', added clarification & removed unnecessary remark

JustinBordage committed:
refactor: Moved schemas into the "sendgrid/definitions" directory

JustinBordage committed:
chore: Replaced the icon with the SendGrid icon

franklevasseur commented on integrations/sendgrid/src/actions/send-mail.ts:
```diff
@@ -0,0 +1,39 @@
+export const sendMail: bp.IntegrationProps['actions']['sendMail'] = async ({ input, logger }) => {
+  try {
+    const requestBody = mapInputToSendMailRequestBody(input)
+    const [response, _] = await sgMail.send(requestBody)
```
**nitpick:**

```ts
await sgMail.send({
    personalizations: [
      {
        to: input.to,
      },
    ],
    from: input.from,
    replyTo: input.replyTo,
    subject: input.subject,
    text: input.body,
  })
```

franklevasseur commented on integrations/sendgrid/src/misc/custom-types.ts:
```diff
@@ -0,0 +1,33 @@
+// ============ Common Types ============
+
+/** A type for modifying the structure of another type */
+export type Modify<T, R> = Omit<T, keyof R> & R
```
**nitpick:** Merge

JustinBordage committed:
refactor: Renamed "Modify" to "Merge" to align with Botpress naming

JustinBordage committed:
refactor: Moved integration name from the package.json to be hardcoded in the integration definition

JustinBordage committed:
chore: Changed the tsconfig to inherit from the repo's tsconfig

JustinBordage committed:
refactor: Inlined "mapInputToSendMailRequestBody" function logic in sendMail action to reduce abstraction

sea-grass commented on integrations/sendgrid/src/misc/utils.ts:
```diff
@@ -0,0 +1,37 @@
+}
+
+export const parseError = (thrown: unknown) => {
+  if (isSendGridError(thrown)) {
```
**suggestion**: `ResponseError` is a class, so you can use `instanceof` to check the value:
```suggestion
  if (thrown instanceof ResponseError) {
```

sea-grass commented on integrations/sendgrid/src/misc/utils.ts:
```diff
@@ -0,0 +1,37 @@
+  if (isSendGridError(thrown)) {
+    const errorMessage = thrown.response.body.errors[0]?.message ?? thrown.message
+    return new RuntimeError(
+      `SendGrid API yielded an error with status code: "${thrown.code}", and message: "${errorMessage}"`,
```
**suggestion**: You can also use [ResponseError's `toString` method](https://github.com/sendgrid/sendgrid-nodejs/blob/main/packages/helpers/classes/response-error.js#L34-L49) to get a nicely formatted error message. (The added bonus is that you don't need to do the in-depth type checking in `isSendGridError` - this method handles the error formatting for you already.)
```suggestion
      `SendGrid API error: ${thrown.toString()}`,
```

sea-grass reviewed (commented):
Looking good so far! Left a couple comments.

JustinBordage committed:
chore: Add "@sendgrid/client" package

JustinBordage committed:
feat: Add validation request to check if provided API key is valid

JustinBordage committed:
fix: Simplified input validation schema for sendMail action

JustinBordage committed:
refactor: Add http status reason phrase to sendMail output for faster debugging

JustinBordage committed:
chore: Updated icon to official brand icon

JustinBordage committed:
chore: Add suppress intelliJ inspection for 'unused' status codes

JustinBordage committed:
chore(integrations/sendgrid): Write hub.md file (Work-In-Progress)

JustinBordage committed:
Added some resources that might be useful for the maintenance of the integration

JustinBordage committed:
chore(sendgrid): fix prettier lint issues in hub.md

JustinBordage committed:
chore(sendgrid): fix linting issues

JustinBordage committed:
fix(sendgrid): Add "secret" flag to the apiKey config value

franklevasseur commented on integrations/sendgrid/definitions/actions.ts:
```diff
@@ -0,0 +1,33 @@
+   *  API version, the response body is empty when the request is successful.
+   *
+   *  Observed: 2025-06-18 */
+  status: StatusCodeSchema.describe('The http status code & reason phrase').title('Request Status Info'),
```
**issue:** same comment as for resend; we shouldnt return the status code, and empty object output schema is fine.

franklevasseur commented on integrations/sendgrid/definitions/common.ts:
```diff
@@ -0,0 +1,26 @@
+/** A string that represents an email address.
+ *
+ *  @remark "correspondent" can refer to both the sender and the receiver of an email. */
+export const EmailAddressSchema = NonBlankString.email().brand('Email')
```
**question:** why the `.brand()` ?

franklevasseur commented on integrations/sendgrid/hub.md:
```diff
@@ -0,0 +1,43 @@
+
+### Enable 2FA
+
+1. TODO
```
**chore:** remove TODOs from the hub.md before releasing

franklevasseur commented on integrations/sendgrid/hub.md:
```diff
@@ -0,0 +1,43 @@
+
+1. TODO
+
+## Actions
```
**chore:** no meed to document your integration actions

franklevasseur commented on integrations/sendgrid/src/actions/send-mail.ts:
```diff
@@ -0,0 +1,29 @@
+export const sendMail: bp.IntegrationProps['actions']['sendMail'] = async ({ input, logger }) => {
+  try {
+    const [response, _] = await sgMail.send({
+      // Transforms input schema data into SendGrid
```
**nitpick:** this comment is not needed; you should trust that your colleagues will understand your intention

franklevasseur commented on integrations/sendgrid/src/index.ts:
```diff
@@ -0,0 +1,33 @@
+      },
+      (error, [resp]) => {
+        if (error) {
+          throw parseError(error)
```
**issue:** as we talked about in person earlier today; make sure that this works

JustinBordage replied:
This has been fixed and tested. The runtime error now reaches Botpress studio when an invalid api key is passed.

franklevasseur commented on integrations/sendgrid/src/misc/HttpStatus.ts:
```diff
@@ -0,0 +1,62 @@
+// noinspection JSUnusedGlobalSymbols
+
+export enum HttpStatus {
```
**nitpick:** same comment as for resend; that's a bit much; your colleagues know what a status 200 is

JustinBordage replied:
The way the HttpStatus was being used was as a means of translating the code into it's reason phrase, whatever the status code may be. The status code that the API responds with any of the code, not just "200 - OK", hence why the full list of codes was added.

Example of how I was using it: 
- `HttpStatus[418] // Returns "IM_A_TEAPOT"`
- `HttpStatus[507] // Returns "INSUFFICIENT_STORAGE"`

That being said I've removed it since the integration no longer returns status codes.

franklevasseur reviewed (commented):
last round of comments before release

JustinBordage committed:
chore: Remove status code return in sendMail output

- This is done to align the integration experience with the rest of the Botpress platform.

JustinBordage committed:
chore: Remove superfluous comment

JustinBordage committed:
chore: Remove unnecessary sections in hub.md

JustinBordage committed:
refactor: Rework the error handling in the API key validator

JustinBordage committed:
chore: remove type branding from email address schema

JustinBordage committed:
chore: Write domain authentication of the hub.md

JustinBordage committed:
chore: fix prettier linter issues

JustinBordage committed:
update: add instanceof check for the "isSendGridError" function

JustinBordage committed:
chore: revert version to v0.0.1 for first official release

franklevasseur committed:
update
